clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashCollection.java

```diff
@@ -284,12 +304,49 @@ public void clear() {
         return listIterator(0);
     }
 
+    @Override
+    public Object[] toArray() {
+        Object[] array = new Object[size];
+        int i = 0;
+        for (Iterator<E> iter = iterator(); iter.hasNext(); )  {
+            array[i++] = iter.next();
+        }
+        return array;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> T[] toArray(T[] array) {
+        if (size > array.length) {
+            array = (T[])java.lang.reflect.Array
+                .newInstance(array.getClass().getComponentType(), size);
+        }
+        int i = 0;
+        for (Iterator<E> iter = iterator(); iter.hasNext(); )  {
+            array[i++] = (T) iter.next();
+        }
+        while (i < array.length) {
+            array[i++] = null;
+        }
+        return array;
+    }
+
     private ListIterator<E> listIterator(int index) {
         return new ImplicitLinkedHashCollectionIterator(index);
     }
 
-    final int slot(Element[] curElements, Object e) {
-        return (e.hashCode() & 0x7fffffff) % curElements.length;
+    final int slot(Element[] curElements, Element e) {
+        return (e.keyHash() & 0x7fffffff) % curElements.length;
+    }
+
+    final private int findIndexOfEqualObject(Object object) {
+        Element value = null;
+        try {
+            value = (Element) object;
+        } catch (ClassCastException e) {
+            return INVALID_INDEX;
+        }
```

Comment:
It's better to do `instanceof` than this. The JVM folds the `instanceof` into the trap it needs to ensure JVM semantics.